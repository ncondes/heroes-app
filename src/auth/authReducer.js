import { types } from '../types/types';

// const state = {
//     name: 'Nicolas',
//     logged: true
// }

// const loginAction = {
//     type: types.login,
//     payload: {
//         name: 'Nicolas',
//         email: 'Nicolas@gmail.com'
//     }
// }

export const authReducer = ( state = {}, action ) => {

    switch ( action.type ) {
        case types.login:
            return {
                ...action.payload,
                logged: true,
            }
        case types.logout:
            return {
                logged: false,
            }
    
        default:
            return state;
    }

}
