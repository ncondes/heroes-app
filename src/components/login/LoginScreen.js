import { useContext } from 'react';
import { useNavigate } from 'react-router-dom';
import { AuthContext } from '../../auth/authContext';
import { types } from '../../types/types';
import { useForm } from '../../hooks/useForm';


export const LoginScreen = () => {

    const [ values, handleInputChange ] = useForm({
        username: '',
    });

    const { dispatch } = useContext( AuthContext );

    const navigate = useNavigate();

    const handleLogin = (e) => {

        e.preventDefault();
        const action = {
            type: types.login,
            payload: {
                name: values.username,
            }
        }

        dispatch( action );

        const lastPath = localStorage.getItem('lastPath') || '/';

        navigate( lastPath, {
            replace: true,
        });

    }

    return (
        <div className='container mt-5'>
            <h1>Login</h1>
            <hr />

            <div className='d-flex justify-content-center'>
                <form className='col-12 col-sm-6' onSubmit={ handleLogin }>
                    <div className='mb-3'>
                        <label
                            htmlFor='username'
                            className='form-label'
                        >
                            Username
                        </label>
                        <input
                            type='text'
                            className='form-control'
                            name='username'
                            id='username'
                            placeholder='Your username here'
                            autoComplete='off'
                            onChange={ handleInputChange }
                        />
                    </div>
                    <div className='d-grid gap-2'>
                        <button
                            type='submit'
                            className='btn btn-outline-dark'
                        >
                            Login
                        </button>
                    </div>
                </form>
            </div>

        </div>
    )
}
