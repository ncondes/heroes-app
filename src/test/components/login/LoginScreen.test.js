import { mount } from 'enzyme';
import { AuthContext } from '../../../auth/authContext';
import { LoginScreen } from '../../../components/login/LoginScreen';
import { MemoryRouter } from 'react-router-dom';


const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate,
}))

describe('Pruebas en LoginScreen', () => {


    const contextValue = {
        dispatch : jest.fn(),
        user: {
            logged: false,
        }
    }

    const wrapper = mount(
        <AuthContext.Provider value={ contextValue }>
            <MemoryRouter initialEntries={ ['/login'] }>
                <LoginScreen />
            </MemoryRouter>
        </AuthContext.Provider>
    );

    test('Debe de hacer match con el snapshot', () => {

        expect( wrapper ).toMatchSnapshot();

    })

    test('Debe de realizar el dispatch y la navegacion', () => {

        wrapper.find('form').prop('onSubmit')({
            preventDefault: () => {}
        });
        expect( contextValue.dispatch ).toHaveBeenCalledWith( {"payload": {"name": ""}, "type": "[auth] Login" });
        expect( mockNavigate ).toHaveBeenCalledWith( "/", {"replace": true});

        localStorage.setItem('lastPath', '/dc');

        wrapper.find('form').prop('onSubmit')({
            preventDefault: () => {}
        });

        expect( mockNavigate ).toHaveBeenCalledWith("/dc", {"replace": true});
        
    })
    
})