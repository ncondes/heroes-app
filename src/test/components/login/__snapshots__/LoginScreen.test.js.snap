// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en LoginScreen Debe de hacer match con el snapshot 1`] = `
<div
  className="container mt-5"
>
  <h1>
    Login
  </h1>
  <hr />
  <div
    className="d-flex justify-content-center"
  >
    <form
      className="col-12 col-sm-6"
      onSubmit={[Function]}
    >
      <div
        className="mb-3"
      >
        <label
          className="form-label"
          htmlFor="username"
        >
          Username
        </label>
        <input
          autoComplete="off"
          className="form-control"
          id="username"
          name="username"
          onChange={[Function]}
          placeholder="Your username here"
          type="text"
        />
      </div>
      <div
        className="d-grid gap-2"
      >
        <button
          className="btn btn-outline-dark"
          type="submit"
        >
          Login
        </button>
      </div>
    </form>
  </div>
</div>
`;
